tensorflow/core/kernels/data/dataset_test_base.cc

```diff
@@ -0,0 +1,182 @@
+  lib_def_ = MakeUnique<FunctionLibraryDefinition>(OpRegistry::Global(), proto);
+
+  OptimizerOptions opts;
+  pflr_ = MakeUnique<ProcessFunctionLibraryRuntime>(
+      device_mgr_.get(), Env::Default(), TF_GRAPH_DEF_VERSION, lib_def_.get(),
+      opts, thread_pool_.get(), nullptr /* cluster_flr */);
+  flr_ = pflr_->GetFLR("/job:localhost/replica:0/task:0/cpu:0");
+  if (thread_pool_ == nullptr) {
+    runner_ = [](std::function<void()> fn) { fn(); };
+  } else {
+    runner_ = [this](std::function<void()> fn) {
+      thread_pool_->Schedule(std::move(fn));
+    };
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::RunOpKernel(OpKernel* op_kernel,
+                                       OpKernelContext* context) {
+  device_->Compute(op_kernel, context);
+  return context->status();
+}
+
+std::unique_ptr<OpKernelContext> DatasetOpsTestBase::CreateOpKernelContext(
+    OpKernel* kernel, gtl::InlinedVector<TensorValue, 4>* inputs) {
+  params_ = MakeUnique<OpKernelContext::Params>();
+  params_->device = device_.get();
+  params_->resource_manager = device_->resource_manager();
+  params_->frame_iter = FrameAndIter(0, 0);
+  params_->inputs = inputs;
+  params_->op_kernel = kernel;
+  params_->function_library = flr_;
+  params_->runner = &runner_;
+  step_container_ = MakeUnique<ScopedStepContainer>(0, [](const string&) {});
+  params_->step_container = step_container_.get();
+  checkpoint::TensorSliceReaderCacheWrapper slice_reader_cache_wrapper;
+  params_->slice_reader_cache = &slice_reader_cache_wrapper;
+  SetOutputAttrs();
+  return MakeUnique<OpKernelContext>(params_.get());
+}
+
+std::unique_ptr<SerializationContext>
+DatasetOpsTestBase::CreateSerializationContext() {
+  SerializationContext::Params params;
+  params.flib_def = lib_def_.get();
+  return MakeUnique<SerializationContext>(params);
+}
+
+Status DatasetOpsTestBase::CheckOpKernelInput(
+    const OpKernel& kernel, const gtl::InlinedVector<TensorValue, 4>& inputs) {
+  if (kernel.input_types().size() != inputs.size()) {
+    return errors::Internal("The number of input elements should be ",
+                            kernel.input_types().size(),
+                            ", but got: ", inputs.size());
+  }
+  return Status::OK();
+}
+
+Tensor* DatasetOpsTestBase::AddDatasetInput(
+    gtl::InlinedVector<TensorValue, 4>* inputs, DataTypeVector input_types,
+    DataType dtype, const TensorShape& shape) {
+  CHECK_GT(input_types.size(), inputs->size())
+      << "Adding more inputs than types.";
+  bool is_ref = IsRefType(input_types[inputs->size()]);
+  Tensor* input = new Tensor(allocator_, dtype, shape);
+  tensors_.push_back(input);
+  if (is_ref) {
+    CHECK_EQ(RemoveRefType(input_types[inputs->size()]), dtype);
+    inputs->push_back({&lock_for_refs_, input});
+  } else {
+    CHECK_EQ(input_types[inputs->size()], dtype);
+    inputs->push_back({nullptr, input});
+  }
+  return input;
+}
+
+void DatasetOpsTestBase::SetOutputAttrs() {
```

Comment:
do we need this to be a method in the API? (could we just inline the logic into the `CreateOpKernelContext` body)